audio/src/range_set/range.rs

```diff
@@ -0,0 +1,305 @@
+use std::{
+    cmp::{self, max, min},
+    fmt,
+};
+
+/// A struct that represents a range between two values. The lower boundary is closed, and the upper
+/// boundary is open. This means that a `Range` from 2 to 5 includes 2, 3 and 4, but not 5.
+/// `std::ops::Range` is not used since it is not `Copy`, which uglifies the resulting code
+/// significantly. For more info, see [this](https://github.com/rust-lang/rust/issues/18045).
+///
+/// ### Invariants
+/// For a range type to be valid, the end must be greater than or equal to the start. For this
+/// reason the following code panics:
+/// ```ignore
+/// let r: Range = (2..0).into();
+/// ```
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct Range {
+    start: usize,
+    len: usize,
+}
+
+impl fmt::Display for Range {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "[{}, {}]", self.start, self.end() - 1)
+    }
+}
+
+impl From<std::ops::Range<usize>> for Range {
+    fn from(r: std::ops::Range<usize>) -> Self {
+        assert!(r.start <= r.end);
+        Self {
+            start: r.start,
+            len: r.end - r.start,
+        }
+    }
```

Comment:
https://doc.rust-lang.org/std/convert/trait.From.html:

> **Note: This trait must not fail.** If the conversion can fail, use `TryFrom`.

And with `TryFrom` it's getting too verbose ... maybe drop this implementation and use the good old `fn new` wherever possible.

Reply:
I will use TryInto, and then add a comment that explains why it will not panic whenever I unwrap it. Does that sound good?

Reply:
I see now that I did not read your comment well, I will look into it tomorrow.

Reply:
@Johannesd3 Alternatively if a range that is given in descending order is constructed, I can reorder `begin` and `end` to get an infallible conversion.

Reply:
I'm not sure. It's just a matter of fact that this range and std's range are not compatible, so it's probably not good to use the syntactic sugar for ranges to create our Range.